File: task3-unknown-language/Swift-notes-app/Sources/App/Handlers/NoteHandler.swift
import Foundation
import Vapor

/// HTTP обработчики для работы с заметками
struct NoteHandler {
    private let noteService: NoteServiceProtocol
    
    init(noteService: NoteServiceProtocol = NoteService()) {
        self.noteService = noteService
    }
    
    /// GET /notes - Получить все заметки
    func getAllNotes(req: Request) async throws -> Response {
        do {
            let notes = try noteService.getAllNotes()
            return try await req.response.json(notes)
        } catch let error as NoteServiceError {
            return try await createErrorResponse(req: req, error: error)
        } catch {
            return try await createErrorResponse(
                req: req,
                status: .internalServerError,
                message: "Внутренняя ошибка сервера"
            )
        }
    }
    
    /// GET /notes/{id} - Получить заметку по ID
    func getNoteById(req: Request) async throws -> Response {
        guard let idString = req.parameters.get("id"),
              let id = Int(idString) else {
            return try await createErrorResponse(
                req: req,
                status: .badRequest,
                message: "Некорректный ID заметки"
            )
        }
        
        do {
            let note = try noteService.getNoteById(id)
            return try await req.response.json(note)
        } catch let error as NoteServiceError {
            return try await createErrorResponse(req: req, error: error)
        } catch {
            return try await createErrorResponse(
                req: req,
                status: .internalServerError,
                message: "Внутренняя ошибка сервера"
            )
        }
    }
    
    /// POST /notes - Создать новую заметку
    func createNote(req: Request) async throws -> Response {
        do {
            let createRequest = try req.content.decode(CreateNoteRequest.self)
            try CreateNoteRequest.validate(content: req)
            
            let note = try noteService.createNote(from: createRequest)
            return try await req.response.json(note, status: .created)
        } catch let error as ValidationError {
            return try await createErrorResponse(
                req: req,
                status: .badRequest,
                message: "Ошибка валидации: \(error.localizedDescription)"
            )
        } catch let error as NoteServiceError {
            return try await createErrorResponse(req: req, error: error)
        } catch {
            return try await createErrorResponse(
                req: req,
                status: .badRequest,
                message: "Некорректные данные запроса"
            )
        }
    }
    
    /// PUT /notes/{id} - Обновить заметку
    func updateNote(req: Request) async throws -> Response {
        guard let idString = req.parameters.get("id"),
              let id = Int(idString) else {
            return try await createErrorResponse(
                req: req,
                status: .badRequest,
                message: "Некорректный ID заметки"
            )
        }
        
        do {
            let updateRequest = try req.content.decode(UpdateNoteRequest.self)
            try UpdateNoteRequest.validate(content: req)
            
            let note = try noteService.updateNote(id, with: updateRequest)
            return try await req.response.json(note)
        } catch let error as ValidationError {
            return try await createErrorResponse(
                req: req,
                status: .badRequest,
                message: "Ошибка валидации: \(error.localizedDescription)"
            )
        } catch let error as NoteServiceError {
            return try await createErrorResponse(req: req, error: error)
        } catch {
            return try await createErrorResponse(
                req: req,
                status: .badRequest,
                message: "Некорректные данные запроса"
            )
        }
    }
    
    /// DELETE /notes/{id} - Удалить заметку
    func deleteNote(req: Request) async throws -> Response {
        guard let idString = req.parameters.get("id"),
              let id = Int(idString) else {
            return try await createErrorResponse(
                req: req,
                status: .badRequest,
                message: "Некорректный ID заметки"
            )
        }
        
        do {
            try noteService.deleteNote(id)
            return Response(status: .noContent)
        } catch let error as NoteServiceError {
            return try await createErrorResponse(req: req, error: error)
        } catch {
            return try await createErrorResponse(
                req: req,
                status: .internalServerError,
                message: "Внутренняя ошибка сервера"
            )
        }
    }
    
    // MARK: - Helper Methods
    
    /// Создать ответ с ошибкой на основе NoteServiceError
    private func createErrorResponse(req: Request, error: NoteServiceError) async throws -> Response {
        let errorResponse = ErrorResponse(
            error: "NoteServiceError",
            message: error.localizedDescription
        )
        return try await req.response.json(errorResponse, status: error.httpStatus)
    }
    
    /// Создать ответ с ошибкой с кастомным статусом и сообщением
    private func createErrorResponse(req: Request, status: HTTPResponseStatus, message: String) async throws -> Response {
        let errorResponse = ErrorResponse(
            error: "RequestError",
            message: message
        )
        return try await req.response.json(errorResponse, status: status)
    }
}

// MARK: - Response Extensions

extension Response {
    /// Создать JSON ответ
    func json<T: Encodable>(_ object: T, status: HTTPResponseStatus = .ok) async throws -> Response {
        self.status = status
        self.headers.contentType = .json
        try self.content.encode(object)
        return self
    }
}

File: task3-unknown-language/Kotlin-notes-app/src/main/kotlin/com/notes/handler/NoteHandler.kt
package com.notes.handler

import com.notes.model.CreateNoteRequest
import com.notes.model.ErrorResponse
import com.notes.model.UpdateNoteRequest
import com.notes.service.NoteNotFoundException
import com.notes.service.NoteService
import com.notes.service.ValidationException
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*

/**
 * HTTP обработчики для работы с заметками
 */
class NoteHandler(private val noteService: NoteService) {
    
    /**
     * Настройка маршрутов для заметок
     */
    fun configureRoutes(routing: Routing) {
        routing {
            route("/notes") {
                // GET /notes - получить все заметки
                get {
                    try {
                        val notes = noteService.getAllNotes()
                        call.respond(HttpStatusCode.OK, notes)
                    } catch (e: Exception) {
                        call.respond(
                            HttpStatusCode.InternalServerError,
                            ErrorResponse("Внутренняя ошибка сервера: ${e.message}")
                        )
                    }
                }
                
                // POST /notes - создать новую заметку
                post {
                    try {
                        val request = call.receive<CreateNoteRequest>()
                        val note = noteService.createNote(request.text)
                        call.respond(HttpStatusCode.Created, note)
                    } catch (e: ValidationException) {
                        call.respond(
                            HttpStatusCode.BadRequest,
                            ErrorResponse(e.message ?: "Ошибка валидации")
                        )
                    } catch (e: Exception) {
                        call.respond(
                            HttpStatusCode.InternalServerError,
                            ErrorResponse("Внутренняя ошибка сервера: ${e.message}")
                        )
                    }
                }
                
                // PUT /notes/{id} - обновить заметку
                put("/{id}") {
                    try {
                        val id = call.parameters["id"]?.toLongOrNull()
                        if (id == null) {
                            call.respond(
                                HttpStatusCode.BadRequest,
                                ErrorResponse("Неверный формат ID")
                            )
                            return@put
                        }
                        
                        val request = call.receive<UpdateNoteRequest>()
                        val note = noteService.updateNote(id, request.text)
                        call.respond(HttpStatusCode.OK, note)
                    } catch (e: ValidationException) {
                        call.respond(
                            HttpStatusCode.BadRequest,
                            ErrorResponse(e.message ?: "Ошибка валидации")
                        )
                    } catch (e: NoteNotFoundException) {
                        call.respond(
                            HttpStatusCode.NotFound,
                            ErrorResponse(e.message ?: "Заметка не найдена")
                        )
                    } catch (e: Exception) {
                        call.respond(
                            HttpStatusCode.InternalServerError,
                            ErrorResponse("Внутренняя ошибка сервера: ${e.message}")
                        )
                    }
                }
                
                // DELETE /notes/{id} - удалить заметку
                delete("/{id}") {
                    try {
                        val id = call.parameters["id"]?.toLongOrNull()
                        if (id == null) {
                            call.respond(
                                HttpStatusCode.BadRequest,
                                ErrorResponse("Неверный формат ID")
                            )
                            return@delete
                        }
                        
                        noteService.deleteNote(id)
                        call.respond(HttpStatusCode.NoContent)
                    } catch (e: NoteNotFoundException) {
                        call.respond(
                            HttpStatusCode.NotFound,
                            ErrorResponse(e.message ?: "Заметка не найдена")
                        )
                    } catch (e: Exception) {
                        call.respond(
                            HttpStatusCode.InternalServerError,
                            ErrorResponse("Внутренняя ошибка сервера: ${e.message}")
                        )
                    }
                }
                
                // GET /notes/{id} - получить заметку по ID (дополнительный эндпоинт)
                get("/{id}") {
                    try {
                        val id = call.parameters["id"]?.toLongOrNull()
                        if (id == null) {
                            call.respond(
                                HttpStatusCode.BadRequest,
                                ErrorResponse("Неверный формат ID")
                            )
                            return@get
                        }
                        
                        val note = noteService.getNoteById(id)
                        call.respond(HttpStatusCode.OK, note)
                    } catch (e: NoteNotFoundException) {
                        call.respond(
                            HttpStatusCode.NotFound,
                            ErrorResponse(e.message ?: "Заметка не найдена")
                        )
                    } catch (e: Exception) {
                        call.respond(
                            HttpStatusCode.InternalServerError,
                            ErrorResponse("Внутренняя ошибка сервера: ${e.message}")
                        )
                    }
                }
            }
        }
    }
}